feat(website): update component page headers

shleewhite opened the pull request:
**Contributing to Twilio**

> All third-party contributors acknowledge that any contributions they provide will be made under the same open-source license that the open-source project is provided under.

- [ ] I acknowledge that all my contributions will be made under the project's license.

andipants12 commented on packages/paste-website/src/pages/layout/flex/index.mdx:
```diff
@@ -59,8 +59,13 @@ export const pageQuery = graphql`
   storybookUrl="/?path=/story/layout-flex--flex-alignment-options"
-  data={props.data.allPasteLayout.edges}
-  packageStatus={props.data.allAirtable.edges}
+  packageName={props.data.allPasteLayout.edges[0].node.name}
```
I think there is an opportunity here to use some lightweight data normalization to simplify the logic in these components. By using the data without normalization we are depending on the shape of the data extracted by the Airtable/graphQL integration, which is subject to change. Additionally, it makes these files dense to read, which can make it easier to introduce bugs. 

Normalization would allow us to contain our logic that extracts these nodes from the graphQL response in a single place. I think it also makes our component APIs more resilient against breaking changes from dependency changes (e.g. if we move away from gatsby).

katieporter commented:
This looks _so_ good! Yay! Just a couple of things that I noticed:
- On components that are still pending peer review (not n/a, but actually pending), there is no "peer review pending" badge (see Alert Dialog as an example).
- /core has a Storybook link and I don't think it's supposed to (or maybe it's just pointing to the wrong URL?).
- Can the description text in the header use `font-size-40`?
- The placement of the Popover on the alpha/beta badge is a little wonky:
![image](https://user-images.githubusercontent.com/5374045/141505400-51457ed5-01fd-4387-ba6c-16b44413a5e1.png)
- Can the icon size and the text line-height in the alpha/beta badge match the "pending" badges so everything is the same height?
- There are some layout adjustments to be made to the header—do we want to tackle those in this ticket or in DSYS-3102 that updates widths and spacing for the entire component page? The latter probably makes more sense but wanted to double check.

shleewhite commented:
> This looks _so_ good! Yay! Just a couple of things that I noticed:
> 
> * On components that are still pending peer review (not n/a, but actually pending), there is no "peer review pending" badge (see Alert Dialog as an example).
> * /core has a Storybook link and I don't think it's supposed to (or maybe it's just pointing to the wrong URL?).
> * Can the description text in the header use `font-size-40`?
> * The placement of the Popover on the alpha/beta badge is a little wonky:
>   ![image](https://user-images.githubusercontent.com/5374045/141505400-51457ed5-01fd-4387-ba6c-16b44413a5e1.png)
> * Can the icon size and the text line-height in the alpha/beta badge match the "pending" badges so everything is the same height?
> * There are some layout adjustments to be made to the header—do we want to tackle those in this ticket or in DSYS-3102 that updates widths and spacing for the entire component page? The latter probably makes more sense but wanted to double check.

@katieporter thanks for looking at it!!
* Ooooh great catch, fixed.
* /core has a link to storybook on the live site, but I can remove it if you don't think its necessary!
* Fixed.
* Yup.. the popover placement is a known bug - it will be fixed when [this PR is merged](https://github.com/twilio-labs/paste/pull/2029)
* Fixed.
* Can you be more specific about which layout adjustments are missing?

katieporter commented:
> > This looks _so_ good! Yay! Just a couple of things that I noticed:
> > 
> > * On components that are still pending peer review (not n/a, but actually pending), there is no "peer review pending" badge (see Alert Dialog as an example).
> > * /core has a Storybook link and I don't think it's supposed to (or maybe it's just pointing to the wrong URL?).
> > * Can the description text in the header use `font-size-40`?
> > * The placement of the Popover on the alpha/beta badge is a little wonky:
> >   ![image](https://user-images.githubusercontent.com/5374045/141505400-51457ed5-01fd-4387-ba6c-16b44413a5e1.png)
> > * Can the icon size and the text line-height in the alpha/beta badge match the "pending" badges so everything is the same height?
> > * There are some layout adjustments to be made to the header—do we want to tackle those in this ticket or in DSYS-3102 that updates widths and spacing for the entire component page? The latter probably makes more sense but wanted to double check.
> 
> @katieporter thanks for looking at it!!
> 
> * Ooooh great catch, fixed.
> * /core has a link to storybook on the live site, but I can remove it if you don't think its necessary!
> * Fixed.
> * Yup.. the popover placement is a known bug - it will be fixed when [this PR is merged](https://github.com/twilio-labs/paste/pull/2029)
> * Fixed.
> * Can you be more specific about which layout adjustments are missing?

- Nice, peer review pending badge looks great!
- Ah, you're right, it does have a link on the live site. It felt weird that it just linked to the first story in our storybook, but maybe that's fine. I trust you and the other engineers to make that call because I am not the target audience. 😂
- I'm still seeing fontSize30 instead of 40.
- Awesome. 👍
- Alpha / beta badge size looks great!
- Yes, I'll DM you the layout adjustments since it'll probably be easier to talk through via Slack.

nkrantz reviewed (commented):
Just found this, I think it's only happening on this page because of the longer heading and having both badges
<img width="926" alt="Screen Shot 2021-11-15 at 2 26 42 PM" src="https://user-images.githubusercontent.com/75342690/141842391-098abe1d-bbf8-45b6-88d0-b4a3c84f89c0.png">

andipants12 commented on packages/paste-website/src/components/shortcodes/component-header/index.tsx:
```diff
@@ -2,160 +2,123 @@ import * as React from 'react';
+    ? `${categoryRoute.replace('/', '')}/${getNameFromPackageName(packageName)}`
+    : undefined;
+
+  const showStatus = status || figmaStatus || designCommitteeReview || engineerCommitteeReview;
```
Personal preference: Would prefer to see this business logic in the component itself. The Status component API includes each of these as a prop on its interface. Since they are known, I think it makes sense to move this conditional into that component. That makes the return body of this component a lot easier to read and more clear.

shleewhite replied:
Updated.

andipants12 reviewed (changes requested):
Great start! I left some feedback about fine tuning the conditional render logic and leveraging the normalization a little more heavily to simplify our components.

loreina reviewed (commented):
This looks absolutely fire 🔥  Just a couple notes:

- Can the badges be placed below the heading for smaller screen widths? See this [Figma spacing spec](https://www.figma.com/file/w8Usm0zhowDJJ9kig5qm3Y/Component-Page-Updates---2021?node-id=136%3A29055)
- Can the description text size be changed to `font-size-40`?

shleewhite commented on packages/paste-website/src/components/badgePopover/BadgePopoverButton.tsx:
```diff
@@ -0,0 +1,31 @@
+    const popover = React.useContext(BadgePopoverContext);
+
+    return (
+      <NonModalDialogDisclosurePrimitive {...popover} ref={ref} as={ButtonBadge} variant={variant}>
```
It doesn't, it expects needs the ref to be passed. This matches how [Popover works](https://github.com/twilio-labs/paste/blob/5fe0831a6c5def59e0381730d3c643bc9eac70d7/packages/paste-core/components/popover/src/PopoverButton.tsx#L24).

shleewhite commented on packages/paste-website/src/components/shortcodes/component-header/index.tsx:
```diff
@@ -2,160 +2,123 @@ import * as React from 'react';
-import {sentenceCase} from '../../../utils/SentenceCase';
-import type {PackageStatusObject} from '../../../utils/types';
+import {STORYBOOK_DOMAIN} from '../../../constants';
+import type {SidebarCategoryRoutes} from '../../../constants';
```
It can't be a type... It's used a bunch of different places as a const. If I import it as a const in this file, I get an eslint error bc it is only used as type.... so I think for now this is the best solution.

shleewhite replied:
Hmm... When I switch to that I still need to do `import type {SidebarCategoryRoutes} from '../../../constants';`

I'm going to leave it as is for now.

shleewhite commented on packages/paste-website/src/components/shortcodes/package-status-legend/index.tsx:
```diff
@@ -1,69 +1,77 @@
-      </Stack>
+    <Box display="flex" alignItems="center" flexWrap="wrap" flexGrow={1} css={{gap: theme.space.space40}}>
+      {status ? <PackageStatusBadge status={status} /> : null}
+      {isFigmaPending ? (
```
Updated the prop to package status, but I'm slightly confused by the other part. 

It only shows a badge if figmaStatus is null, if figmaStatus is a string it shouldnt show anything. So, if I make figmaStatus = undefined at the normalization stage, I'd need to check that a prop was not passed like this:
`!figmaStatus ? <Badge>...</Badge> : null` . Is that better?

shleewhite replied:
ok! I updated the normalizer where if the value is null, it doesnt add the key/value to the normalized data.

shleewhite commented on packages/paste-website/src/pages/components/alert/index.mdx:
```diff
@@ -62,13 +62,11 @@ export const pageQuery = graphql`
   storybookUrl="/?path=/story/components-alert--all-variant"
-  data={props.data.allPasteComponent.edges}
-  packageStatus={props.data.allAirtable.edges}
+  githubUrl="https://github.com/twilio-labs/paste/tree/main/packages/paste-core/components/alert"
```
Hmmm... For now, can it stay on the mdx pages? Right now, the normalizer handles all these:

```
    case 'allPasteComponent':
    case 'allPasteDesignTokensPackage':
    case 'allPasteLayout':
    case 'allPasteLibraries':
    case 'allPastePattern':
    case 'allPastePrimitive':
    case 'allPasteThemePackage':
```

tokensPackage, libraries, patterns, and  themePackage have different or no github URLs. 

This ticket has increased in scope a couple of times, so I would definitely prefer to improve/finalize the normalization function to another ticket (maybe this one: https://issues.corp.twilio.com/browse/DSYS-3105 ?).

andipants12 replied:
Yeah that totally makes sense! I forgot for a moment that _wasn't actually part of the AC for this work._ Do you mind creating a ticket?

shleewhite commented:
Phew! I think I addressed all the comments:
* Description font size is _for sure_ `fontSize40` now
* Badges wrap below the heading if the heading + badges are wider than `space70`
* BadgePopover works now the badge ref fix is in
* `getNormalizedHeaderData` returns undefined if the value from graphql is null

@loreina @nkrantz @andipants12

andipants12 commented on packages/paste-website/src/components/shortcodes/package-status-legend/index.tsx:
```diff
@@ -1,70 +1,83 @@
+    popoverContent = StatusDescriptions.BETA;
+  }
 
   return (
```
Unless the outer fragments are required, could we remove them? I find that functional components are clearest when written most like functions.

```suggestion
if (badgeVariant != null) {
    return (
        <BadgePopoverContainer baseId="status">
            <BadgePopoverButton variant={badgeVariant}>
              <>
                <NewIcon decorative size="sizeIcon10" />
                {status}
              </>
            </BadgePopoverButton>
            <BadgePopover aria-label={status}>{popoverContent}</BadgePopover>
        </BadgePopoverContainer>
    )
}

return null;
```

Also, is the inner fragment required because the button wrapper expects a single child? If so, I think it's worth a comment explaining so folks don't remove it in the future thinking it's superfluous. 

If it's not required👇 

```suggestion
if (badgeVariant != null) {
    return (
        <BadgePopoverContainer baseId="status">
            <BadgePopoverButton variant={badgeVariant}>
                <NewIcon decorative size="sizeIcon10" />
                {status}
            </BadgePopoverButton>
            <BadgePopover aria-label={status}>{popoverContent}</BadgePopover>
        </BadgePopoverContainer>
    )
}

return null;
```

shleewhite replied:
Updated.

andipants12 reviewed (commented):
Responded to the comments/discussion. Biggest take away is yes, lets def create a separate ticket for polishing up the normalization stuff. Thanks for adding it in to this effort (I know it added quite a bit of scope)!

shleewhite commented on packages/paste-website/src/components/shortcodes/package-status-legend/index.tsx:
```diff
@@ -1,70 +1,83 @@
-  } = packageStatus[0].node.data;
+const PackageStatusBadge: React.FC<{status: string}> = ({status}) => {
+  let badgeVariant: BadgeVariants | undefined;
+  let popoverContent = '';
```
Updated.

loreina reviewed (approved):
header lookin' great, no 🧢  (get it bc.. HEADer)

zahnster reviewed (approved):
So good lets 🚢  it
